File: Easy/RectangleArea.java
/*
Find the total area covered by two rectilinear rectangles in a 2D plane.
Each rectangle is defined by its bottom left corner and top right corner as shown in the figure.
Assume that the total area is never beyond the maximum possible value of int.
*/

public class Solution {
    public int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
        int areaOfSqrA = (C - A) * (D - B);
        int areaOfSqrB = (G - E) * (H - F);

        int left = Math.max(A,E);
        int right = Math.min(C,G);
        int top = Math.min(D,H);
        int bottom = Math.max(B,F);

        int overlap = 0; // count for overlap area
        if (right > left && top > bottom) overlap = (right - left) * (top - bottom);
        return areaOfSqrA + areaOfSqrB - overlap;
    }
}

/*
Reference:
https://leetcode.com/discuss/39398/my-java-solution-sum-of-areas-overlapped-area
*/

File: Easy/RectangleArea.py
"""
Find the total area covered by two rectilinear rectangles in a 2D plane.
Each rectangle is defined by its bottom left corner and top right corner as shown in the figure.
Assume that the total area is never beyond the maximum possible value of int.
"""

class Solution(object):
    def computeArea(self, A, B, C, D, E, F, G, H):
        """
        :type A: int
        :type B: int
        :type C: int
        :type D: int
        :type E: int
        :type F: int
        :type G: int
        :type H: int
        :rtype: int
        """
        areaOfSqrA = (C - A) * (D - B)
        areaOfSqrB = (G - E) * (H - F)

        left = max(A,E)
        right = min(C,G)
        top = min(D,H)
        bottom = max(B,F)

        if (right > left and top > bottom):
            return areaOfSqrA + areaOfSqrB - (right - left) * (top - bottom)
        else:
            return areaOfSqrA + areaOfSqrB
